commit 537cd89484ab57ca38ae25d9557361c0815977d1
Author: Barnabás Pőcze <pobrn@protonmail.com>
Date:   Tue Dec 15 20:47:10 2020 -0800

    fault-injection: handle EI_ETYPE_TRUE
    
    Commit af3b854492f351d1 ("mm/page_alloc.c: allow error injection")
    introduced EI_ETYPE_TRUE, but did not extend
    
     * lib/error-inject.c:error_type_string(), and
     * kernel/fail_function.c:adjust_error_retval()
    
    to accommodate for this change.
    
    Handle EI_ETYPE_TRUE in both functions appropriately by
     * returning "TRUE" in error_type_string(),
     * adjusting the return value to true (1) in adjust_error_retval().
    
    Furthermore, simplify the logic of handling EI_ETYPE_NULL in
    adjust_error_retval().
    
    Link: https://lkml.kernel.org/r/njB1czX0ZgWPR9h61euHIBb5bEyePw9D4D2m3i5lc9Cl96P8Q1308dTcmsEZW7Vtz3Ifz4do-rOtSfuFTyGoEDYokkK2aUqBePVptzZEWfU=@protonmail.com
    Signed-off-by: Barnabás Pőcze <pobrn@protonmail.com>
    Acked-by: Masami Hiramatsu <mhiramat@kernel.org>
    Reviewed-by: Akinobu Mita <akinobu.mita@gmail.com>
    Cc: "Naveen N. Rao" <naveen.n.rao@linux.ibm.com>
    Cc: Anil S Keshavamurthy <anil.s.keshavamurthy@intel.com>
    Cc: "David S. Miller" <davem@davemloft.net>
    Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>

diff --git a/kernel/fail_function.c b/kernel/fail_function.c
index b0b1ad93fa95..60dc825ecc2b 100644
--- a/kernel/fail_function.c
+++ b/kernel/fail_function.c
@@ -37,9 +37,7 @@ static unsigned long adjust_error_retval(unsigned long addr, unsigned long retv)
 {
 	switch (get_injectable_error_type(addr)) {
 	case EI_ETYPE_NULL:
-		if (retv != 0)
-			return 0;
-		break;
+		return 0;
 	case EI_ETYPE_ERRNO:
 		if (retv < (unsigned long)-MAX_ERRNO)
 			return (unsigned long)-EINVAL;
@@ -48,6 +46,8 @@ static unsigned long adjust_error_retval(unsigned long addr, unsigned long retv)
 		if (retv != 0 && retv < (unsigned long)-MAX_ERRNO)
 			return (unsigned long)-EINVAL;
 		break;
+	case EI_ETYPE_TRUE:
+		return 1;
 	}
 
 	return retv;